server/src/main/java/org/elasticsearch/index/mapper/BlockDocValuesReader.java

```diff
@@ -408,10 +408,10 @@ private static class SingletonDoubles extends BlockDocValuesReader {
         }
 
         @Override
-        public BlockLoader.Block read(BlockFactory factory, Docs docs) throws IOException {
+        public BlockLoader.Block read(BlockFactory factory, Docs docs, int offset) throws IOException {
             try (BlockLoader.DoubleBuilder builder = factory.doublesFromDocValues(docs.count())) {
                 int lastDoc = -1;
-                for (int i = 0; i < docs.count(); i++) {
+                for (int i = offset; i < docs.count(); i++) {
```

Comment:
I think we should subtract the offset when creating the builder. For example, it should be `factory.doublesFromDocValues(docs.count() - offset)` instead.

Reply:
I've pushed a commit with this.